File: app/src/main/java/com/worldventures/dreamtrips/wallet/analytics/BaseCardDetailsAction.java
package com.worldventures.dreamtrips.wallet.analytics;

import com.worldventures.dreamtrips.core.utils.tracksystem.Attribute;
import com.worldventures.dreamtrips.wallet.domain.entity.card.BankCard;

public abstract class BaseCardDetailsAction extends WalletAnalyticsAction {

   @Attribute("paycardtype") String paycardType = "Unknown";
   @Attribute("paycardissuer") String paycardIssuer = "Unknown";
   @Attribute("cardtype") final String cardtype = "Payment";

   public void fillPaycardInfo(BankCard bankCard) {
      if (bankCard.issuerInfo().financialService() == null) return;
      switch (bankCard.issuerInfo().financialService()) {
         case AMEX:
            paycardIssuer = "American Express";
            paycardType = "American Express";
            break;
         case VISA:
            paycardIssuer = bankCard.issuerInfo().bankName();
            paycardType = "Visa";
            break;
         case MASTERCARD:
            paycardIssuer = bankCard.issuerInfo().bankName();
            paycardType = "MasterCard";
            break;
         case DISCOVER:
            paycardIssuer = "Discover";
            paycardType = "Discover";
            break;
      }
   }
}

File: wallet/src/main/java/com/worldventures/wallet/analytics/BaseCardDetailsAction.kt
package com.worldventures.wallet.analytics

import com.worldventures.core.service.analytics.AttributeMap
import com.worldventures.wallet.domain.entity.record.FinancialService
import com.worldventures.wallet.domain.entity.record.Record
import java.util.HashMap

abstract class BaseCardDetailsAction : WalletAnalyticsAction() {

   @AttributeMap protected val attributeMap: MutableMap<String, String> = HashMap()

   open fun fillRecordDetails(record: Record) {

      val cardType = "Payment"
      val paymentCardType: String
      val paymentCardIssuer: String

      when (record.financialService) {
         FinancialService.AMEX -> {
            paymentCardIssuer = "American Express"
            paymentCardType = "American Express"
         }
         FinancialService.VISA -> {
            paymentCardIssuer = record.bankName
            paymentCardType = "Visa"
         }
         FinancialService.MASTERCARD -> {
            paymentCardIssuer = record.bankName
            paymentCardType = "MasterCard"
         }
         FinancialService.DISCOVER -> {
            paymentCardIssuer = "Discover"
            paymentCardType = "Discover"
         }
         else -> {
            paymentCardIssuer = "Unknown"
            paymentCardType = "Unknown"
         }
      }

      attributeMap.put("paycardtype", paymentCardType)
      attributeMap.put("paycardissuer", paymentCardIssuer)
      attributeMap.put("cardtype", cardType)
   }
}
